In [1]:
def top_set_sym_freq_cnt(top_set_syms_n_freq):
    # accommodate upto 5 periods of days_lookbacks(i.e. days_lookbacks = [5, 10, 15, 20, 25])
    sym_freq_cnt_15 = []
    sym_freq_cnt_14 = []
    sym_freq_cnt_13 = []
    sym_freq_cnt_12 = []
    sym_freq_cnt_11 = []
    sym_freq_cnt_10 = []
    sym_freq_cnt_9 = []
    sym_freq_cnt_8 = []
    sym_freq_cnt_7 = []
    sym_freq_cnt_6 = []
    sym_freq_cnt_5 = []
    sym_freq_cnt_4 = []
    sym_freq_cnt_3 = []
    sym_freq_cnt_2 = []

    for sym_n_freq in top_set_syms_n_freq:
        _sym = sym_n_freq[0]
        _freq = sym_n_freq[1]
        if _freq == 15:
            sym_freq_cnt_15.append(_sym)
        elif _freq == 14:
            sym_freq_cnt_14.append(_sym)
        elif _freq == 13:
            sym_freq_cnt_13.append(_sym)
        elif _freq == 12:
            sym_freq_cnt_12.append(_sym)                        
        elif _freq == 11:
            sym_freq_cnt_11.append(_sym)
        elif _freq == 10:
            sym_freq_cnt_10.append(_sym)            
        elif _freq == 9:
            sym_freq_cnt_9.append(_sym)
        elif _freq == 8:
            sym_freq_cnt_8.append(_sym)
        elif _freq == 7:
            sym_freq_cnt_7.append(_sym)  
        elif _freq == 6:
            sym_freq_cnt_6.append(_sym)
        elif _freq == 5:
            sym_freq_cnt_5.append(_sym)
        elif _freq == 4:
            sym_freq_cnt_4.append(_sym)
        elif _freq == 3:
            sym_freq_cnt_3.append(_sym)          
        else:
            sym_freq_cnt_2.append(_sym)

    l_sym_freq_cnt = []

    l_sym_freq_cnt.append(sym_freq_cnt_15)
    l_sym_freq_cnt.append(sym_freq_cnt_14)
    l_sym_freq_cnt.append(sym_freq_cnt_13)
    l_sym_freq_cnt.append(sym_freq_cnt_12)    
    l_sym_freq_cnt.append(sym_freq_cnt_11)   
    l_sym_freq_cnt.append(sym_freq_cnt_10)
    l_sym_freq_cnt.append(sym_freq_cnt_9)
    l_sym_freq_cnt.append(sym_freq_cnt_8)
    l_sym_freq_cnt.append(sym_freq_cnt_7)    
    l_sym_freq_cnt.append(sym_freq_cnt_6)
    l_sym_freq_cnt.append(sym_freq_cnt_5)
    l_sym_freq_cnt.append(sym_freq_cnt_4)
    l_sym_freq_cnt.append(sym_freq_cnt_3)    
    l_sym_freq_cnt.append(sym_freq_cnt_2)    

    return l_sym_freq_cnt    

In [2]:
import pandas as pd
# import numpy as np
import datetime
# from IPython.display import display, HTML
from yf_utils import _2_split_train_val_test, _3_random_slices, _4_lookback_slices
from yf_utils import _5_perf_ranks, _6_grp_tuples_sort_sum
from myUtils import pickle_load, pickle_dump

pd.set_option('display.max_rows', 500)
pd.set_option('display.max_columns', 30)
pd.set_option('display.max_colwidth', 16)
pd.set_option('display.width', 790)

path_dir = "C:/Users/ping/MyDrive/stocks/yfinance/"
path_data_dump = path_dir + "VSCode_dump/"

fp_df_close_clean = 'df_close_clean'




# fp_df_eval_results = f'df_eval_results'
# fp_df_picks = f'df_picks'



###########################################################################################################
# str_today = datetime.date.today().strftime('%Y-%m-%d')
# # str_today = '2023-03-08'
# # plan to run tran, validate and test periodically, so add date to file name
# fp_df_eval_sym_freq_results_train    = f'df_eval_sym_freq_results_train_{str_today}'
# fp_df_eval_sym_freq_results_validate = f'df_eval_sym_freq_results_validate_{str_today}'
# fp_df_eval_sym_freq_results_test     = f'df_eval_sym_freq_results_test_{str_today}'
# # plan to accumulate current results, therefore no date added to file name
# fp_df_picks  = f'df_picks'

# fp_df_eval_sym_freq_results_train    = f'df_eval_sym_freq_results_train'
# fp_df_eval_sym_freq_results_validate = f'df_eval_sym_freq_results_validate'
# fp_df_eval_sym_freq_results_test     = f'df_eval_sym_freq_results_test'
# fp_df_picks  = f'df_picks'
###########################################################################################################

# df_close_clean = pickle_load(path_data_dump, fp_df_close_clean)

#### Check run results

In [3]:
# df = pickle_load(path_data_dump, fp_df_eval_sym_freq_results_train)
# df
# df = pickle_load(path_data_dump, fp_df_eval_sym_freq_results_validate)
# df
# df = pickle_load(path_data_dump, fp_df_eval_sym_freq_results_test)
# df
# df = pickle_load(path_data_dump, fp_df_picks)
# df

#### Create column names to store current picks

In [4]:
# # columns for current picks
# col_picks0 = ['date_end_df_train', 'max_days_lookbacks', 'days_lookbacks']
# col_picks1 = ['sym_freq_15', 'sym_freq_14', 'sym_freq_13', 'sym_freq_12', 'sym_freq_11', 'sym_freq_10', 'sym_freq_9']
# col_picks2 = ['sym_freq_8',  'sym_freq_7',  'sym_freq_6',  'sym_freq_5',  'sym_freq_4',  'sym_freq_3',  'sym_freq_2']
# col_picks_total = col_picks0 + col_picks1 + col_picks2
# print(f'col_picks_total:\n{col_picks_total}, total columns: {len(col_picks_total)}')

#### DANGER: Create a new Dataframe to Picks. Previous Picks will be destroyed.

In [5]:
# # # Creating Empty DataFrame and save it to file
# df = pd.DataFrame(columns=col_picks_total)
# pickle_dump(df, path_data_dump, fp_df_picks)
# df

#### Create column names to store run results

In [6]:
# col_add0 = ['n_samples', 'days_lookbacks', 'days_eval', 'n_top_syms', 'syms_start', 'syms_end', 'sym_freq_cnt']
# col_add1 = ['grp(CAGR)_mean',         'grp(CAGR)_std',         'grp(CAGR)_mean/std']
# col_add2 = ['grp(CAGR/UI)_mean',      'grp(CAGR/UI)_std',      'grp(CAGR/UI)_mean/std']
# col_add3 = ['grp(CAGR/retnStd)_mean', 'grp(CAGR/retnStd)_std', 'grp(CAGR/retnStd)_mean/std']
# col_add4 = ['grp(retnStd/UI)_mean',   'grp(retnStd/UI)_std',   'grp(retnStd/UI)_mean/std']
# col_add5 = ['SPY_CAGR', 'SPY_CAGR/UI', 'SPY_CAGR/retnStd', 'SPY_retnStd/UI']

# col_add_total = col_add0 + col_add1 + col_add2 + col_add3 + col_add4 + col_add5
# print(f'col_add_total:\n{col_add_total}, total columns: {len(col_add_total)}')

#### DANGER: Create a new Dataframe to store run results. Previous data will be destroyed. To create a new dataframe, uncomment one choice from below.

In [7]:
# # # Creating Empty DataFrame and save it to file
# df = pd.DataFrame(columns=col_add_total)
# pickle_dump(df, path_data_dump, fp_df_eval_results)
# df

In [8]:
# # # Creating Empty DataFrame and save it to file
# df = pd.DataFrame(columns=col_add_total)
# pickle_dump(df, path_data_dump, fp_df_eval_sym_freq_results_train)
# df

# # # Creating Empty DataFrame and save it to file
# df = pd.DataFrame(columns=col_add_total)
# pickle_dump(df, path_data_dump, fp_df_eval_sym_freq_results_validate)
# df

# # # Creating Empty DataFrame and save it to file
# df = pd.DataFrame(columns=col_add_total)
# pickle_dump(df, path_data_dump, fp_df_eval_sym_freq_results_test)
# df

# # # Creating Empty DataFrame and save it to file
# df = pd.DataFrame(columns=col_add_total)
# pickle_dump(df, path_data_dump, fp_df_picks)
# df

#### Select run parameters. Parameters can also be passed using papermill by running yf_7_freq_cnt_pm_.ipynb

In [9]:
fp_df_eval_results = f'df_eval_results'
fp_df_picks = f'df_picks'


# SELECT RUN PARAMETERS.async Parameters can also be passed using papermill by running yf_7_freq_cnt_pm_.ipynb
verbose = True  # True prints more output
# verbose = False  # True prints more output

# write run results to df_eval_results
# store_results = False
store_results = True

# select run type
# run_type = 'train'
# run_type = 'validate'
# run_type = 'test'
run_type = 'current'

# number of max lookback tuples to create for iloc iloc_start_train:iloc_end_train:iloc_end_eval
# i.e. number of grp_top_set_syms_n_freq and grp_top_set_syms 
# n_samples = 400  
n_samples = 40

# for training, the number of days to lookback from iloc max-lookback iloc_end_train
# days_lookbacks = [30, 60, 120]
days_lookbacks = [15, 30, 60, 120]
days_lookbacks.sort()

# number of days from iloc_end_train are used to evaluate effectiveness of the training
days_eval = 4
# days_eval = 5

# number of the most-common symbols from days_lookbacks' performance rankings to keep
n_top_syms = 20  

syms_start = 0  #  start index of n_top_syms for evaluation
syms_end = 10  #  end index of n_top_syms for evaluation

# over-ride parameters for run_type 'current'
if run_type == 'current':
  days_eval = 0  # no need to eval, df_eval will be empty
  n_samples = 1  # no need to repeat sample the current result, repeat sample will yield the same tuple

print(f'verbose : {verbose }')
print(f'store_results: {store_results}')
print(f'run_type: {run_type}')
print(f'n_samples: {n_samples}')
print(f'days_lookbacks: {days_lookbacks}')
print(f'days_eval: {days_eval}')
print(f'n_top_syms: {n_top_syms}')
print(f'syms_start: {syms_start}')
print(f'syms_end: {syms_end}')

verbose : True
store_results: True
run_type: current
n_samples: 1
days_lookbacks: [15, 30, 60, 120]
days_eval: 0
n_top_syms: 20
syms_start: 0
syms_end: 10


In [10]:
# Parameters
fp_df_eval_results = "df_eval_results"
fp_df_picks = "df_picks"
verbose = True
store_results = True
n_top_syms = 20
syms_start = 0
syms_end = 10
run_type = "test"
n_samples = 6
days_eval = 5
days_lookbacks = [15, 30, 60, 120]


#### Load previous run results

In [11]:
df_eval_results = pickle_load(path_data_dump, fp_df_eval_results)
df_picks = pickle_load(path_data_dump, fp_df_picks)
df_close_clean = pickle_load(path_data_dump, fp_df_close_clean)

#### Drop the last n row in df_close_clean to get previous picks

In [12]:
# # drop the last n row(s)
# n = 1
# df_close_clean = df_close_clean[:-n]
# df_close_clean.tail(2)

#### Split dataframe into Train, Validate and Test

In [13]:
# Split df_close_clean into training (df_train), validation (df_val) and test (df_test) set.
# The default split is 0.7, 0.2, 0.1 respectively.
df_train, df_val, df_test = _2_split_train_val_test(df_close_clean)

#### Create df_current, a df with the latest data

In [14]:
max_days_lookbacks = max(days_lookbacks)
print(f'max_days_lookbacks: {max_days_lookbacks}')
df_current = df_test.tail(max_days_lookbacks)  
df_current.tail(3)

max_days_lookbacks: 120


A     AA    AAL         AAP        AAPL        AAWW         AB        ABB        ABBV         ABC        ABM    ABR        ABT       ACGL       ACHC  ...       XRAY    XRX        XYL       YELP       YUMC         YY         ZBH        ZBRA         ZD         ZG       ZION        ZTO         ZTS   ZUMZ        ZWS
Date                                                                                                                                                                       ...                                                                                                                                                                
2023-03-14  138.399994  44.59  14.66  121.699997  152.589996   99.300003  35.709999  33.759998  153.850006  149.289993  44.570000  12.12  98.550003  67.339996  69.160004  ...  38.130001  15.14  99.900002  29.260000  60.680000  27.469999  126.629997  292.459991  74.370003  39.610001  31.309999  26.270000  164.559998  20.43  21.379999
2023-03-15  134.029999  39.77  13.86  121.769997  152.990005  102.449997  35.080002  31.990000  154.059998  149.610001  44.250000  11.53  97.800003  62.740002  68.459999  ...  37.209999  15.00  96.550003  29.170000  60.439999  27.070000  125.260002  287.739990  73.779999  39.919998  30.709999  25.940001  163.570007  19.50  20.980000
2023-03-16  136.820007  40.75  14.12  122.029999  155.850006  102.459999  35.310001  32.549999  155.300003  152.050003  44.790001  11.50  99.029999  66.070000  70.230003  ...  37.439999  15.20  97.160004  29.780001  60.980000  28.000000  125.430000  294.929993  74.510002  40.849998  32.110001  28.000000  166.309998  19.25  20.969999

[3 rows x 1290 columns]

#### Load df according to run_type

In [15]:
if run_type == 'train':
  # df_eval_sym_freq_results_train = pickle_load(path_data_dump, fp_df_eval_sym_freq_results_train)
  df = df_train.copy()    
elif run_type == 'validate':
  # df_eval_sym_freq_results_validate = pickle_load(path_data_dump, fp_df_eval_sym_freq_results_validate)
  df = df_val.copy()
elif run_type == 'test':
  # df_eval_sym_freq_results_test = pickle_load(path_data_dump, fp_df_eval_sym_freq_results_test)
  df = df_test.copy()
elif run_type == 'current':  # get the current results
  # df_picks = pickle_load(path_data_dump, fp_df_picks)  
  df = df_current.copy()
else:
  msg_stop = f"ERROR run_type must be 'train', 'validate', 'test' or 'current', run_type is: {run_type}"
  raise SystemExit(msg_stop )  

#### Print dataframe for the run, and lengths of other dataframes

In [16]:
print(f'run_type: {run_type}, df.tail(3):\n{df.tail(3)}\n')
len_df = len(df)
len_df_train = len(df_train)
len_df_val = len(df_val)
len_df_test = len(df_test)
len_df_current = len(df_current)
print(f'run_type: {run_type}, len(df): {len(df)}')
print(f'len_df_train: {len_df_train}, len_df_val: {len_df_val}, len_df_test: {len_df_test}, len_df_current: {len_df_current} ')


run_type: test, df.tail(3):
                     A     AA    AAL         AAP        AAPL        AAWW         AB        ABB        ABBV         ABC        ABM    ABR        ABT       ACGL       ACHC  ...       XRAY    XRX        XYL       YELP       YUMC         YY         ZBH        ZBRA         ZD         ZG       ZION        ZTO         ZTS   ZUMZ        ZWS
Date                                                                                                                                                                       ...                                                                                                                                                                
2023-03-14  138.399994  44.59  14.66  121.699997  152.589996   99.300003  35.709999  33.759998  153.850006  149.289993  44.570000  12.12  98.550003  67.339996  69.160004  ...  38.130001  15.14  99.900002  29.260000  60.680000  27.469999  126.629997  292.459991  74.370003  39.610001  31.309999  26.2700

#### Create a sets of iloc lookback slices (iloc_start_train:iloc_end_train:iloc_end_eval), where:  
* iloc_end_train - iloc_start_train = days_lookback
* iloc_end_eval - iloc_end_train = days_eval  
#### for example, if given:  
* n_samples = 2  
* days_lookbacks = [10, 20, 30]  
* days_eval = 5  
#### a possible result is:  
  - max_lookback_slices: [(417, 447, 452), (265, 295, 300)], where:
    - len(max_lookback_slices) = n_samples = 2
    - middle number in the tuples, 447 and 295, is the iloc of the "pivot day" for the days in "days_lookbacks" to lookback
    - 447 - 417 = middle number - first number = max(days_lookbacks) = 30
    - 295 - 265 = middle number - first number = max(days_lookbacks) = 30      
    - 452 - 447 = last number - middle number = days_eval = 5
    - 300 - 295 = last number - middle number = days_eval = 5    
  - sets_lookback_slices: [[(437, 447, 452), (427, 447, 452), (417, 447, 452)], [(285, 295, 300), (275, 295, 300), (265, 295, 300)]], where:
    - len(sets_lookback_slices) = n_samples = 2
    - last tuple in each list, i.e. (417, 447, 452) and (265, 295, 300), is a tuple from max_lookback_slices
    - where a set, e.g. [(437, 447, 452), (427, 447, 452), (417, 447, 452)]:
      - middle number, 447, iloc of the "pivot day" is constant for that set
      - middle number - first number, is the training period specified in days_lookbacks
        - 447 - 437 = middle number - first number = days_lookbacks[0] = 10
        - 447 - 427 = middle number - first number = days_lookbacks[1] = 20
        - 447 - 417 = middle number - first number = days_lookbacks[2] = 30
      - last number, 452, iloc of the end of the evaluation period is constant
        - 452 - 447 = last number - middle number = days_eval = 5

In [17]:
# return n_samples slices
max_lookback_slices = _3_random_slices(len_df, n_samples=n_samples, days_lookback=max(days_lookbacks), days_eval=days_eval, verbose=False)
# return n_samples * len(days_lookbacks) slices
sets_lookback_slices = _4_lookback_slices(max_slices=max_lookback_slices, days_lookbacks=days_lookbacks, verbose=False)

if verbose:
  print(f'number of max_lookback_slices is equal to n_samples = {n_samples}')
  print(f'max_lookback_slices:\n{max_lookback_slices}\n')
  print(f'number of sets in sets_lookback_slices is equal to n_samples = {n_samples}')
  print(f'days_lookbacks: {days_lookbacks}')  
  print(f'sets_lookback_slices:\n{sets_lookback_slices}\n')
  print(f'number of tuples in each "set of lookback slices" is equal to len(days_lookbacks): len({days_lookbacks})')    

number of max_lookback_slices is equal to n_samples = 6
max_lookback_slices:
[(17, 137, 142), (18, 138, 143), (3, 123, 128), (18, 138, 143), (7, 127, 132), (20, 140, 145)]

number of sets in sets_lookback_slices is equal to n_samples = 6
days_lookbacks: [15, 30, 60, 120]
sets_lookback_slices:
[[(122, 137, 142), (107, 137, 142), (77, 137, 142), (17, 137, 142)], [(123, 138, 143), (108, 138, 143), (78, 138, 143), (18, 138, 143)], [(108, 123, 128), (93, 123, 128), (63, 123, 128), (3, 123, 128)], [(123, 138, 143), (108, 138, 143), (78, 138, 143), (18, 138, 143)], [(112, 127, 132), (97, 127, 132), (67, 127, 132), (7, 127, 132)], [(125, 140, 145), (110, 140, 145), (80, 140, 145), (20, 140, 145)]]

number of tuples in each "set of lookback slices" is equal to len(days_lookbacks): len([15, 30, 60, 120])


#### Generate grp_top_set_syms_n_freq. It is a list of sub-lists, e.g.:
 - [[('AGY', 7), ('PCG', 7), ('KDN', 6), ..., ('CYT', 3)], ..., [('FCN', 9), ('HIG', 9), ('SJR', 8), ..., ('BFH', 2)]]
#### grp_top_set_syms_n_freq has n_samples sub-lists. Each sub-list corresponds to a tuple in the max_lookback_slices. Each sub-list has n_top_syms tuples of (symbol, frequency) pairs, and is sorted in descending order of frequency. The frequency is the number of times the symbol appears in the top n_top_syms performance rankings of CAGR/UI, CAGR/retnStd and retnStd/UI.
#### Therefore, symbols in the sub-list are the best performing symbols for the periods in days_lookbacks. Each sub-list corresponds to a tuple in max_lookback_slices. There are as many sub-lists as there are tuples in max_lookback_slices.

In [18]:
# grp_top_set_syms_n_freq is a list of lists of top_set_syms_n_freq, e.g.
#   [[('AGY', 7), ('PCG', 7), ('KDN', 6), ..., ('CYT', 3)],
#    [('FCN', 9), ('HIG', 9), ('SJR', 8), ..., ('BFH', 2)]]
#   where each list is the best performing symbols from a lb_slices, e.g.
#     [(483, 513, 523), (453, 513, 523), (393, 513, 523)]  
grp_top_set_syms_n_freq = []  # list of lists of top_set_symbols_n_freq, there are n_samples lists in list
grp_top_set_syms = []  # grp_top_set_syms_n_freq without the frequency count

# lb_slices, e.g  [(483, 513, 523), (453, 513, 523), (393, 513, 523)],
#  is one max_lookback_slice, e.g. (393, 513, 523), along with
#  the remaining slices of the days_lookbacks, e.g. (483, 513, 523), (453, 513, 523)  
for i, lb_slices in enumerate(sets_lookback_slices):
  print(f'\n########## {i + 1} of {len(sets_lookback_slices)} lb_slices in sets_lookcak_slices ##########')
  # unsorted list of the most frequent symbols in performance metrics of the lb_slices  
  grp_most_freq_syms = []
  for j, lb_slice in enumerate(lb_slices):  # lb_slice, e.g. (246, 276, 286)
    iloc_start_train = lb_slice[0]     # iloc of start of training period
    iloc_end_train   = lb_slice[1]     # iloc of end of training period
    iloc_start_eval  = iloc_end_train  # iloc of start of evaluation period
    iloc_end_eval    = lb_slice[2]     # iloc of end of evaluation period
    lookback         = iloc_end_train - iloc_start_train
    d_eval           = iloc_end_eval - iloc_start_eval

    _df = df.iloc[iloc_start_train:iloc_end_train]
    date_start_df_train = _df.index[0].strftime('%Y-%m-%d')
    date_end_df_train = _df.index[-1].strftime('%Y-%m-%d')

    if verbose:
      print(f'days lookback:       {lookback},  {j + 1} of {len(days_lookbacks)} days_lookbacks: {days_lookbacks}')
      print(f'lb_slices:           {lb_slices}')
      print(f'lb_slice:            {lb_slice}')
      print(f'days eval:           {d_eval}')    
      print(f'iloc_start_train:    {iloc_start_train}')
      print(f'iloc_end_train:      {iloc_end_train}')
      print(f'date_start_df_train: {date_start_df_train}')
      print(f'date_end_df_train:   {date_end_df_train}')


    perf_ranks, most_freq_syms = _5_perf_ranks(_df, n_top_syms=n_top_syms)
    # unsorted list of the most frequent symbols in performance metrics of the lb_slices  
    grp_most_freq_syms.append(most_freq_syms)  
    if verbose:    
      # 1 lookback of r_CAGR/UI, r_CAGR/retnStd, r_retnStd/UI
      print(f'perf_ranks: {perf_ranks}')  
      # most common symbols of perf_ranks 
      print(f'most_freq_syms: {most_freq_syms}')     
      # grp_perf_ranks[lookback] = perf_ranks
      print(f'+++ finish lookback slice {lookback} +++\n')

  if verbose:
    print(f'grp_most_freq_syms: {grp_most_freq_syms}')
    # grp_most_freq_syms a is list of lists of tuples of 
    #  the most-common-symbols symbol:frequency cumulated from
    #  each days_lookback  
    print(f'**** finish lookback slices {lb_slices} ****\n')

  # flatten list of lists of (symbol:frequency)
  flat_grp_most_freq_syms = [val for sublist in grp_most_freq_syms for val in sublist]
  # return "symbol, frequency" pairs of the most frequent symbols, i.e. best performing symbols,
  #  in flat_grp_most_freq_syms. The paris are sorted in descending frequency.   
  set_most_freq_syms = _6_grp_tuples_sort_sum(flat_grp_most_freq_syms, reverse=True)
  # get the top n_top_syms of the most frequent "symbol, frequency" pairs
  top_set_syms_n_freq = set_most_freq_syms[0:n_top_syms]
  # get symbols from top_set_syms_n_freq, i[0] = symbol, i[1]=symbol's frequency count
  top_set_syms = [i[0] for i in top_set_syms_n_freq[syms_start:syms_end]]  

  # grp_top_set_syms_n_freq is a list of lists of top_set_syms_n_freq, e.g.
  #   [[('AGY', 7), ('PCG', 7), ('KDN', 6), ..., ('CYT', 3)],
  #    [('FCN', 9), ('HIG', 9), ('SJR', 8), ..., ('BFH', 2)]]
  #   where each list is the best performing symbols from a lb_slices, e.g.
  #     [(483, 513, 523), (453, 513, 523), (393, 513, 523)]    
  grp_top_set_syms_n_freq.append(top_set_syms_n_freq)
  grp_top_set_syms.append(top_set_syms)

  if verbose:  
    print(f'top {n_top_syms} ranked symbols and frequency from set {lb_slices}:\n{top_set_syms_n_freq}')
    print(f'top {n_top_syms} ranked symbols from set {lb_slices}:\n{top_set_syms}')  
    print(f'===== finish top {n_top_syms} ranked symbols from days_lookback set {lb_slices} =====\n\n')


########## 1 of 6 lb_slices in sets_lookcak_slices ##########
days lookback:       15,  1 of 4 days_lookbacks: [15, 30, 60, 120]
lb_slices:           [(122, 137, 142), (107, 137, 142), (77, 137, 142), (17, 137, 142)]
lb_slice:            (122, 137, 142)
days eval:           5
iloc_start_train:    122
iloc_end_train:      137
date_start_df_train: 2023-02-06
date_end_df_train:   2023-02-27


perf_ranks: {'period-15': {'r_CAGR/UI': array(['TA', 'SGEN', 'TNP', 'TNK', 'OEC', 'ACGL', 'LNTH', 'FTI', 'ASC',
       'STNG', 'BP', 'WST', 'PBF', 'CNA', 'AMCX', 'EME', 'NVGS', 'MSGS',
       'PDFS', 'FTSM'], dtype=object), 'r_CAGR/retnStd': array(['TA', 'TNP', 'SGEN', 'TNK', 'STNG', 'ASC', 'FTI', 'OEC', 'ACGL',
       'BP', 'LNTH', 'AMCX', 'PBF', 'WST', 'IRTC', 'HLF', 'FTNT', 'XPRO',
       'NSP', 'HZO'], dtype=object), 'r_retnStd/UI': array(['TA', 'SGEN', 'CNA', 'LAUR', 'OEC', 'KBR', 'TNK', 'LNTH', 'MSGS',
       'FTSM', 'ACGL', 'TWNK', 'QNST', 'EME', 'MMS', 'NVGS', 'TNP',
       'PDFS', 'RELX', 'FCN'], dtype=object)}}
most_freq_syms: [('TA', 3), ('SGEN', 3), ('TNP', 3), ('TNK', 3), ('OEC', 3), ('ACGL', 3), ('LNTH', 3), ('FTI', 2), ('ASC', 2), ('STNG', 2), ('BP', 2), ('WST', 2), ('PBF', 2), ('CNA', 2), ('AMCX', 2), ('EME', 2), ('NVGS', 2), ('MSGS', 2), ('PDFS', 2), ('FTSM', 2), ('IRTC', 1), ('HLF', 1), ('FTNT', 1), ('XPRO', 1), ('NSP', 1), ('HZO', 1), ('LAUR', 1), ('KBR', 1), ('TWNK'

perf_ranks: {'period-30': {'r_CAGR/UI': array(['TA', 'TSLA', 'ELF', 'TNK', 'OEC', 'SGEN', 'LNTH', 'CTLT', 'FTSM',
       'NVDA', 'SMG', 'NVGS', 'LRN', 'ASC', 'PDFS', 'TNP', 'WIRE', 'EURN',
       'THR', 'ACLS'], dtype=object), 'r_CAGR/retnStd': array(['TSLA', 'TA', 'TNK', 'ELF', 'TNP', 'NVDA', 'SGEN', 'SMG', 'LNTH',
       'OEC', 'CTLT', 'ASC', 'CNK', 'WIRE', 'ACLS', 'PAG', 'RVNC', 'EURN',
       'AN', 'THR'], dtype=object), 'r_retnStd/UI': array(['TA', 'ELF', 'OEC', 'FTSM', 'LRN', 'SGEN', 'NVGS', 'LNTH', 'PDFS',
       'TNK', 'AQN', 'MMS', 'CTLT', 'CME', 'KEX', 'STLA', 'EURN', 'FISV',
       'TSLA', 'QNST'], dtype=object)}}
most_freq_syms: [('TA', 3), ('TSLA', 3), ('ELF', 3), ('TNK', 3), ('OEC', 3), ('SGEN', 3), ('LNTH', 3), ('CTLT', 3), ('EURN', 3), ('FTSM', 2), ('NVDA', 2), ('SMG', 2), ('NVGS', 2), ('LRN', 2), ('ASC', 2), ('PDFS', 2), ('TNP', 2), ('WIRE', 2), ('THR', 2), ('ACLS', 2), ('CNK', 1), ('PAG', 1), ('RVNC', 1), ('AN', 1), ('AQN', 1), ('MMS', 1), ('CME', 1), ('KEX', 1), ('ST

perf_ranks: {'period-120': {'r_CAGR/UI': array(['FTSM', 'ELF', 'HZNP', 'RMBS', 'ACGL', 'ANF', 'FTI', 'AGYS', 'TPX',
       'OI', 'LOCO', 'ACLS', 'FIVE', 'HEES', 'CNO', 'LW', 'ASC', 'ATKR',
       'TMHC', 'TEX'], dtype=object), 'r_CAGR/retnStd': array(['FTSM', 'RMBS', 'ACLS', 'ELF', 'FTI', 'TEX', 'HEES', 'ASC', 'TPX',
       'THR', 'ACGL', 'HZNP', 'AXON', 'WYNN', 'ANF', 'LOCO', 'OI', 'FMX',
       'PERI', 'URI'], dtype=object), 'r_retnStd/UI': array(['FTSM', 'HZNP', 'ELF', 'LW', 'ANF', 'AGYS', 'ACGL', 'QNST', 'TMHC',
       'FIVE', 'BIIB', 'REGN', 'ORI', 'OI', 'CNO', 'IRDM', 'LOCO', 'DHI',
       'MMS', 'RMBS'], dtype=object)}}
most_freq_syms: [('FTSM', 3), ('ELF', 3), ('HZNP', 3), ('RMBS', 3), ('ACGL', 3), ('ANF', 3), ('OI', 3), ('LOCO', 3), ('FTI', 2), ('AGYS', 2), ('TPX', 2), ('ACLS', 2), ('FIVE', 2), ('HEES', 2), ('CNO', 2), ('LW', 2), ('ASC', 2), ('TMHC', 2), ('TEX', 2), ('ATKR', 1), ('THR', 1), ('AXON', 1), ('WYNN', 1), ('FMX', 1), ('PERI', 1), ('URI', 1), ('QNST', 1), ('BIIB', 1)

perf_ranks: {'period-60': {'r_CAGR/UI': array(['FTSM', 'BVH', 'ACLS', 'SGEN', 'HSBC', 'MCFT', 'HEES', 'ATI', 'TA',
       'TX', 'URI', 'CRUS', 'THR', 'ALGN', 'RVNC', 'SBH', 'WYNN', 'BLDR',
       'OI', 'TPX'], dtype=object), 'r_CAGR/retnStd': array(['ACLS', 'FTSM', 'BVH', 'TX', 'SGEN', 'CRUS', 'HSBC', 'SMG', 'OEC',
       'BMA', 'TPX', 'ALGN', 'BLDR', 'THR', 'ATI', 'SBH', 'URI', 'HEES',
       'META', 'TEX'], dtype=object), 'r_retnStd/UI': array(['FTSM', 'TA', 'MCFT', 'SGEN', 'LW', 'RVNC', 'HEES', 'OTTR', 'PEN',
       'URI', 'MBI', 'MMS', 'VRNS', 'ACGL', 'WYNN', 'ATI', 'HSBC', 'BVH',
       'NVO', 'AJRD'], dtype=object)}}
most_freq_syms: [('FTSM', 3), ('BVH', 3), ('SGEN', 3), ('HSBC', 3), ('HEES', 3), ('ATI', 3), ('URI', 3), ('ACLS', 2), ('MCFT', 2), ('TA', 2), ('TX', 2), ('CRUS', 2), ('THR', 2), ('ALGN', 2), ('RVNC', 2), ('SBH', 2), ('WYNN', 2), ('BLDR', 2), ('TPX', 2), ('OI', 1), ('SMG', 1), ('OEC', 1), ('BMA', 1), ('META', 1), ('TEX', 1), ('LW', 1), ('OTTR', 1), ('PEN', 1), ('MBI',

perf_ranks: {'period-15': {'r_CAGR/UI': array(['TSLA', 'W', 'META', 'CTLT', 'ALGN', 'JWN', 'LRN', 'ELF', 'ICPT',
       'AN', 'CRUS', 'PAG', 'SMG', 'SHOP', 'CACC', 'SAIA', 'DBRG', 'ALLY',
       'PACW', 'AX'], dtype=object), 'r_CAGR/retnStd': array(['TSLA', 'W', 'CTLT', 'ALGN', 'META', 'JWN', 'SMG', 'ICPT', 'ELF',
       'SHOP', 'AN', 'CACC', 'LRN', 'DBRG', 'SAIA', 'PACW', 'RVNC', 'PRG',
       'PAG', 'CRUS'], dtype=object), 'r_retnStd/UI': array(['LRN', 'META', 'CRUS', 'PAG', 'ELF', 'SUN', 'ALGN', 'OI', 'AX',
       'CUBE', 'ICPT', 'ALLY', 'JWN', 'ST', 'AN', 'OEC', 'JOE', 'BYD',
       'CTLT', 'MRCY'], dtype=object)}}
most_freq_syms: [('META', 3), ('CTLT', 3), ('ALGN', 3), ('JWN', 3), ('LRN', 3), ('ELF', 3), ('ICPT', 3), ('AN', 3), ('CRUS', 3), ('PAG', 3), ('TSLA', 2), ('W', 2), ('SMG', 2), ('SHOP', 2), ('CACC', 2), ('SAIA', 2), ('DBRG', 2), ('ALLY', 2), ('PACW', 2), ('AX', 2), ('RVNC', 1), ('PRG', 1), ('SUN', 1), ('OI', 1), ('CUBE', 1), ('ST', 1), ('OEC', 1), ('JOE', 1), ('BYD', 1), 

perf_ranks: {'period-120': {'r_CAGR/UI': array(['FTSM', 'ELF', 'HZNP', 'COLL', 'RNR', 'AGYS', 'FTI', 'OI', 'SLB',
       'PERI', 'ACGL', 'LOCO', 'XPRO', 'BKE', 'RMBS', 'GILD', 'ASC',
       'LVS', 'NFLX', 'ATKR'], dtype=object), 'r_CAGR/retnStd': array(['FTSM', 'ELF', 'RMBS', 'RNR', 'HZNP', 'XPRO', 'PERI', 'FTI',
       'COLL', 'ASC', 'OI', 'AGYS', 'BMA', 'ATKR', 'PEN', 'ACLS', 'AXON',
       'BKE', 'SLB', 'LOCO'], dtype=object), 'r_retnStd/UI': array(['FTSM', 'HZNP', 'ELF', 'DCP', 'BIIB', 'IRDM', 'ACGL', 'COLL',
       'ISEE', 'AGYS', 'MGI', 'PCG', 'TS', 'NFLX', 'FTI', 'REGN', 'OI',
       'LW', 'VALE', 'SLB'], dtype=object)}}
most_freq_syms: [('FTSM', 3), ('ELF', 3), ('HZNP', 3), ('COLL', 3), ('AGYS', 3), ('FTI', 3), ('OI', 3), ('SLB', 3), ('RNR', 2), ('PERI', 2), ('ACGL', 2), ('LOCO', 2), ('XPRO', 2), ('BKE', 2), ('RMBS', 2), ('ASC', 2), ('NFLX', 2), ('ATKR', 2), ('GILD', 1), ('LVS', 1), ('BMA', 1), ('PEN', 1), ('ACLS', 1), ('AXON', 1), ('DCP', 1), ('BIIB', 1), ('IRDM', 1), ('ISEE',

perf_ranks: {'period-30': {'r_CAGR/UI': array(['TA', 'TNK', 'ELF', 'TSLA', 'SGEN', 'OEC', 'CTLT', 'LNTH', 'LRN',
       'FTSM', 'TEX', 'NVGS', 'WIRE', 'PDFS', 'SMG', 'ASC', 'NVDA',
       'ACLS', 'PKE', 'WST'], dtype=object), 'r_CAGR/retnStd': array(['TA', 'TSLA', 'TNK', 'ELF', 'SGEN', 'CTLT', 'OEC', 'LNTH', 'WIRE',
       'TEX', 'SMG', 'TNP', 'PAG', 'ASC', 'ACLS', 'NVDA', 'PDFS', 'PRG',
       'NVGS', 'LRN'], dtype=object), 'r_retnStd/UI': array(['TA', 'ELF', 'OEC', 'LRN', 'FTSM', 'SGEN', 'NVGS', 'PDFS', 'LNTH',
       'TNK', 'PKE', 'TEX', 'CTLT', 'KEX', 'EME', 'MMS', 'ORI', 'EURN',
       'MWA', 'CME'], dtype=object)}}
most_freq_syms: [('TA', 3), ('TNK', 3), ('ELF', 3), ('SGEN', 3), ('OEC', 3), ('CTLT', 3), ('LNTH', 3), ('LRN', 3), ('TEX', 3), ('NVGS', 3), ('PDFS', 3), ('TSLA', 2), ('FTSM', 2), ('WIRE', 2), ('SMG', 2), ('ASC', 2), ('NVDA', 2), ('ACLS', 2), ('PKE', 2), ('WST', 1), ('TNP', 1), ('PAG', 1), ('PRG', 1), ('KEX', 1), ('EME', 1), ('MMS', 1), ('ORI', 1), ('EURN', 1), ('MWA', 

perf_ranks: {'period-120': {'r_CAGR/UI': array(['FTSM', 'ELF', 'RMBS', 'HZNP', 'ACGL', 'ANF', 'FTI', 'AGYS',
       'ACLS', 'OI', 'HEES', 'ASC', 'TEX', 'CNO', 'LOCO', 'ATKR', 'TMHC',
       'TPX', 'JBL', 'LW'], dtype=object), 'r_CAGR/retnStd': array(['FTSM', 'ACLS', 'RMBS', 'ELF', 'TEX', 'ASC', 'HEES', 'FTI', 'ACGL',
       'WYNN', 'THR', 'AXON', 'ANF', 'HZNP', 'TNK', 'URI', 'ATKR', 'GE',
       'OI', 'AGYS'], dtype=object), 'r_retnStd/UI': array(['FTSM', 'HZNP', 'ELF', 'LW', 'ANF', 'AGYS', 'ACGL', 'QNST', 'TMHC',
       'FIVE', 'IRDM', 'BIIB', 'REGN', 'ORI', 'CNO', 'OI', 'MMS', 'RMBS',
       'LOCO', 'FTI'], dtype=object)}}
most_freq_syms: [('FTSM', 3), ('ELF', 3), ('RMBS', 3), ('HZNP', 3), ('ACGL', 3), ('ANF', 3), ('FTI', 3), ('AGYS', 3), ('OI', 3), ('ACLS', 2), ('HEES', 2), ('ASC', 2), ('TEX', 2), ('CNO', 2), ('LOCO', 2), ('ATKR', 2), ('TMHC', 2), ('LW', 2), ('TPX', 1), ('JBL', 1), ('WYNN', 1), ('THR', 1), ('AXON', 1), ('TNK', 1), ('URI', 1), ('GE', 1), ('QNST', 1), ('FIVE', 1), ('I

perf_ranks: {'period-30': {'r_CAGR/UI': array(['ICPT', 'ACLS', 'CTLT', 'RVNC', 'CRUS', 'NATI', 'SMG', 'TSLA',
       'RCL', 'EVRI', 'MEOH', 'NVDA', 'WYNN', 'FORM', 'META', 'THR',
       'WBD', 'TRIP', 'MGM', 'ALGN'], dtype=object), 'r_CAGR/retnStd': array(['ICPT', 'ACLS', 'TSLA', 'SMG', 'RVNC', 'MARA', 'CTLT', 'WBD',
       'CRUS', 'RCL', 'NVDA', 'TRIP', 'NATI', 'W', 'THR', 'MEOH', 'FORM',
       'EVRI', 'ALGN', 'STAA'], dtype=object), 'r_retnStd/UI': array(['DCP', 'NATI', 'CRUS', 'CTLT', 'FTSM', 'EVRI', 'LRN', 'SWKS',
       'FISV', 'RVNC', 'BYD', 'MEOH', 'AAWW', 'EFR', 'MGM', 'SJR', 'WYNN',
       'URI', 'STX', 'META'], dtype=object)}}
most_freq_syms: [('CTLT', 3), ('RVNC', 3), ('CRUS', 3), ('NATI', 3), ('EVRI', 3), ('MEOH', 3), ('ICPT', 2), ('ACLS', 2), ('SMG', 2), ('TSLA', 2), ('RCL', 2), ('NVDA', 2), ('WYNN', 2), ('FORM', 2), ('META', 2), ('THR', 2), ('WBD', 2), ('TRIP', 2), ('MGM', 2), ('ALGN', 2), ('MARA', 1), ('W', 1), ('STAA', 1), ('DCP', 1), ('FTSM', 1), ('LRN', 1), ('SWKS', 

perf_ranks: {'period-120': {'r_CAGR/UI': array(['FTSM', 'ELF', 'HZNP', 'FTI', 'FIVE', 'ACGL', 'AGYS', 'RNR', 'RCL',
       'OI', 'RMBS', 'COLL', 'GILD', 'TPX', 'LOCO', 'SLB', 'WYNN', 'NFLX',
       'PERI', 'CNO'], dtype=object), 'r_CAGR/retnStd': array(['FTSM', 'ELF', 'RMBS', 'RCL', 'HZNP', 'FTI', 'WYNN', 'RNR', 'PEN',
       'ACLS', 'THR', 'XPRO', 'TEX', 'PERI', 'ATKR', 'OI', 'AGYS', 'SLB',
       'FIVE', 'GILD'], dtype=object), 'r_retnStd/UI': array(['FTSM', 'HZNP', 'ELF', 'BIIB', 'FIVE', 'ACGL', 'AGYS', 'REGN',
       'ISEE', 'NFLX', 'IRDM', 'LOCO', 'OI', 'COLL', 'TPX', 'CNO', 'FTI',
       'PCG', 'MGI', 'GILD'], dtype=object)}}
most_freq_syms: [('FTSM', 3), ('ELF', 3), ('HZNP', 3), ('FTI', 3), ('FIVE', 3), ('AGYS', 3), ('OI', 3), ('GILD', 3), ('ACGL', 2), ('RNR', 2), ('RCL', 2), ('RMBS', 2), ('COLL', 2), ('TPX', 2), ('LOCO', 2), ('SLB', 2), ('WYNN', 2), ('NFLX', 2), ('PERI', 2), ('CNO', 2), ('PEN', 1), ('ACLS', 1), ('THR', 1), ('XPRO', 1), ('TEX', 1), ('ATKR', 1), ('BIIB', 1), ('RE

perf_ranks: {'period-30': {'r_CAGR/UI': array(['TA', 'TNK', 'ELF', 'OEC', 'SGEN', 'TSLA', 'TEX', 'CTLT', 'SMG',
       'ASC', 'AMPH', 'LNTH', 'WIRE', 'THR', 'ASTE', 'LRN', 'PDFS',
       'NTRA', 'NVGS', 'NVDA'], dtype=object), 'r_CAGR/retnStd': array(['TNK', 'TA', 'TSLA', 'ELF', 'SMG', 'OEC', 'CTLT', 'HY', 'TEX',
       'ASC', 'WIRE', 'THR', 'NTRA', 'ASTE', 'SGEN', 'NVDA', 'STRL',
       'ACLS', 'TWLO', 'LNTH'], dtype=object), 'r_retnStd/UI': array(['TA', 'ELF', 'OEC', 'AMPH', 'LRN', 'KBR', 'FTSM', 'SGEN', 'PGR',
       'NVGS', 'LNTH', 'PDFS', 'EME', 'TNK', 'CLX', 'PRGS', 'PKE', 'MWA',
       'AIR', 'MMS'], dtype=object)}}
most_freq_syms: [('TA', 3), ('TNK', 3), ('ELF', 3), ('OEC', 3), ('SGEN', 3), ('LNTH', 3), ('TSLA', 2), ('TEX', 2), ('CTLT', 2), ('SMG', 2), ('ASC', 2), ('AMPH', 2), ('WIRE', 2), ('THR', 2), ('ASTE', 2), ('LRN', 2), ('PDFS', 2), ('NTRA', 2), ('NVGS', 2), ('NVDA', 2), ('HY', 1), ('STRL', 1), ('ACLS', 1), ('TWLO', 1), ('KBR', 1), ('FTSM', 1), ('PGR', 1), ('EME', 1), ('C

perf_ranks: {'period-120': {'r_CAGR/UI': array(['FTSM', 'ELF', 'RMBS', 'ACGL', 'HZNP', 'AGYS', 'ANF', 'FTI',
       'MCFT', 'ACLS', 'OI', 'ASC', 'AXON', 'ATKR', 'HEES', 'LW', 'LOCO',
       'TEX', 'TPX', 'STRL'], dtype=object), 'r_CAGR/retnStd': array(['FTSM', 'RMBS', 'ACLS', 'ELF', 'AXON', 'ASC', 'TEX', 'THR', 'HEES',
       'ACGL', 'MCFT', 'WYNN', 'FTI', 'STRL', 'ATKR', 'AGYS', 'GE', 'FMX',
       'OI', 'ANF'], dtype=object), 'r_retnStd/UI': array(['FTSM', 'HZNP', 'ELF', 'LW', 'AGYS', 'ANF', 'ACGL', 'QNST', 'TMHC',
       'FIVE', 'IRDM', 'BIIB', 'FHN', 'NAVI', 'ORI', 'OI', 'CNO', 'MMS',
       'RMBS', 'FTI'], dtype=object)}}
most_freq_syms: [('FTSM', 3), ('ELF', 3), ('RMBS', 3), ('ACGL', 3), ('AGYS', 3), ('ANF', 3), ('FTI', 3), ('OI', 3), ('HZNP', 2), ('MCFT', 2), ('ACLS', 2), ('ASC', 2), ('AXON', 2), ('ATKR', 2), ('HEES', 2), ('LW', 2), ('TEX', 2), ('STRL', 2), ('LOCO', 1), ('TPX', 1), ('THR', 1), ('WYNN', 1), ('GE', 1), ('FMX', 1), ('QNST', 1), ('TMHC', 1), ('FIVE', 1), ('IRDM', 1)

#### Output the best performing symbols for each set in sets_lookback_slices

In [19]:
for i, top_set_syms_n_freq in enumerate(grp_top_set_syms_n_freq):
  l_sym_freq_cnt = top_set_sym_freq_cnt(top_set_syms_n_freq)
  if verbose:
    print(f'set_lookback_slices: {sets_lookback_slices[i]}')
    print(f'max_lookback_slices: {max_lookback_slices}\n')


    # print(f'data below will be added to {fp_df_picks}')


    print(f'date_end_df_train:   {date_end_df_train}')    
    print(f'max_days_lookbacks:  {max_days_lookbacks}')   
    print(f'days_lookbacks:      {days_lookbacks}')
    print(f'sym_freq_15:         {l_sym_freq_cnt[0]}')
    print(f'sym_freq_14:         {l_sym_freq_cnt[1]}')
    print(f'sym_freq_13:         {l_sym_freq_cnt[2]}')
    print(f'sym_freq_12:         {l_sym_freq_cnt[3]}')
    print(f'sym_freq_11:         {l_sym_freq_cnt[4]}')
    print(f'sym_freq_10:         {l_sym_freq_cnt[5]}')
    print(f'sym_freq_9:          {l_sym_freq_cnt[6]}')
    print(f'sym_freq_8:          {l_sym_freq_cnt[7]}')
    print(f'sym_freq_7:          {l_sym_freq_cnt[8]}')
    print(f'sym_freq_6:          {l_sym_freq_cnt[9]}')
    print(f'sym_freq_5:          {l_sym_freq_cnt[10]}')
    print(f'sym_freq_4:          {l_sym_freq_cnt[11]}')
    print(f'sym_freq_3:          {l_sym_freq_cnt[12]}')
    print(f'sym_freq_2:          {l_sym_freq_cnt[13]}\n')

if run_type == 'current':  # record results to df
  row_picks0      = [date_end_df_train, max_days_lookbacks, str(days_lookbacks)]
  row_picks1      = [str(l_sym_freq_cnt[0]),  str(l_sym_freq_cnt[1]), str(l_sym_freq_cnt[2]),  str(l_sym_freq_cnt[3])]
  row_picks2      = [str(l_sym_freq_cnt[4]),  str(l_sym_freq_cnt[5]), str(l_sym_freq_cnt[6]),  str(l_sym_freq_cnt[7])]
  row_picks3      = [str(l_sym_freq_cnt[8]),  str(l_sym_freq_cnt[9]), str(l_sym_freq_cnt[10]), str(l_sym_freq_cnt[11])]
  row_picks4      = [str(l_sym_freq_cnt[12]), str(l_sym_freq_cnt[13])]
  row_picks_total = row_picks0 + row_picks1 + row_picks2 + row_picks3 + row_picks4
  print(f'row_picks_total: {row_picks_total}')

  df_picks.loc[len(df_picks)] = row_picks_total
  pickle_dump(df_picks, path_data_dump, fp_df_picks)
  print(f'appended row_picks_total to df_picks:\n{row_picks_total}\n')

set_lookback_slices: [(122, 137, 142), (107, 137, 142), (77, 137, 142), (17, 137, 142)]
max_lookback_slices: [(17, 137, 142), (18, 138, 143), (3, 123, 128), (18, 138, 143), (7, 127, 132), (20, 140, 145)]

date_end_df_train:   2023-03-02
max_days_lookbacks:  120
days_lookbacks:      [15, 30, 60, 120]
sym_freq_15:         []
sym_freq_14:         []
sym_freq_13:         []
sym_freq_12:         []
sym_freq_11:         []
sym_freq_10:         ['FTSM']
sym_freq_9:          ['SGEN']
sym_freq_8:          ['TA']
sym_freq_7:          ['ACGL', 'OEC']
sym_freq_6:          ['ACLS', 'ASC', 'ELF', 'LNTH', 'TNK']
sym_freq_5:          ['THR', 'TNP']
sym_freq_4:          ['FTI', 'HEES', 'MMS', 'NVGS', 'PDFS', 'TPX', 'URI']
sym_freq_3:          ['ANF']
sym_freq_2:          []

set_lookback_slices: [(123, 138, 143), (108, 138, 143), (78, 138, 143), (18, 138, 143)]
max_lookback_slices: [(17, 137, 142), (18, 138, 143), (3, 123, 128), (18, 138, 143), (7, 127, 132), (20, 140, 145)]

date_end_df_train:   2023-

In [20]:
# # TODO make this another notebook 
# if run_type == 'current':
#   _df = pickle_load(path_data_dump, fp_df_picks)

# # _df = _df.loc[(_df['days_lookbacks'] == '[30, 60, 120]'), 'sym_freq_8']
# _df = _df.loc[(_df['days_lookbacks'] == '[15, 30, 60, 120]'), 'sym_freq_7']
# # _df = _df.loc[(_df['days_lookbacks'] == '[30, 60, 120]'), 'sym_freq_9']
# _df.values

#### Compare performance of symbols in set_lookback_slices versus SPY

In [21]:
from myUtils import symb_perf_stats_vectorized_v8

print('z_grp_top_set_syms_n_freq:')
# zip max_lookback_slices with top performing symbols:freq pairs
z_grp_top_set_syms_n_freq = zip(max_lookback_slices, grp_top_set_syms_n_freq)

idx_train = pd.Index([])
idx_eval = pd.Index([])
for i, (_lookback_slice, _top_set_syms_n_freq) in enumerate(z_grp_top_set_syms_n_freq):
  iloc_start_train = _lookback_slice[0]
  iloc_end_train = _lookback_slice[1]
  iloc_start_eval = iloc_end_train
  iloc_end_eval = _lookback_slice[2]

  print(f'{i + 1 } of {n_samples} max_lookback_slice: {_lookback_slice}\n')
  # print(f'max_lookback_slice: {_lookback_slice}\n')
  # dates correspond to max_lookback_slice
  idx_train = df.index[iloc_start_train:iloc_end_train]
  date_start_df_train = idx_train[0].strftime('%Y-%m-%d')
  date_end_df_train = idx_train[-1].strftime('%Y-%m-%d')
  print(f'len(idx_train): {len(idx_train)}')
  print(f'idx_train: {idx_train}')
  print(f'df_train_dates (inclusive): {date_start_df_train} - {date_end_df_train}\n')     

  if run_type != 'current':
    idx_eval = df.index[iloc_start_eval:iloc_end_eval]
    date_start_df_eval = idx_eval[0].strftime('%Y-%m-%d')
    date_end_df_eval = idx_eval[-1].strftime('%Y-%m-%d')
    print(f'len(idx_eval): {len(idx_eval)}')
    print(f'idx_eval: {idx_eval}')
    print(f'df_eval dates (inclusive): {date_start_df_eval} - {date_end_df_eval}\n')
  else: # run_type == "current'
    print(f'run_type: {run_type}, idx_eval is empty: {idx_eval.empty}')

  print(f'top_set_syms_n_freq: {_top_set_syms_n_freq}\n')   

  l_sym_freq_cnt = top_set_sym_freq_cnt(_top_set_syms_n_freq)
  if verbose:
    print(f'set_lookback_slices: {sets_lookback_slices[i]}')    
    print(f'max_lookback_slice:  {_lookback_slice}')
    print(f'days_lookbacks:      {days_lookbacks}')
    print(f'date_end_df_train:   {date_end_df_train}')
    print(f'sym_freq_15:         {l_sym_freq_cnt[0]}')
    print(f'sym_freq_14:         {l_sym_freq_cnt[1]}')
    print(f'sym_freq_13:         {l_sym_freq_cnt[2]}')
    print(f'sym_freq_12:         {l_sym_freq_cnt[4]}')
    print(f'sym_freq_10:         {l_sym_freq_cnt[5]}')
    print(f'sym_freq_9:          {l_sym_freq_cnt[6]}')
    print(f'sym_freq_8:          {l_sym_freq_cnt[7]}')
    print(f'sym_freq_7:          {l_sym_freq_cnt[8]}')
    print(f'sym_freq_6:          {l_sym_freq_cnt[9]}')
    print(f'sym_freq_5:          {l_sym_freq_cnt[10]}')
    print(f'sym_freq_4:          {l_sym_freq_cnt[11]}')
    print(f'sym_freq_3:          {l_sym_freq_cnt[12]}')
    print(f'sym_freq_2:          {l_sym_freq_cnt[13]}\n') 

  if idx_eval.empty:
    if run_type == 'current':
      msg_stop = f'\n\n**** Normal termination for run_type "{run_type}" ****'
      raise SystemExit(msg_stop)
    else:
      msg_stop = f'\n\n**** STOPPED df_eval is empty but run_type != "current", the run_type is "{run_type}" ****'
      raise SystemExit(msg_stop)    

  _sym_idx = ['SPY']
  df_SPY = df[iloc_start_eval:iloc_end_eval][_sym_idx]

  (
    _symbols,
    _period_yr,
    _retn,
    _DD,
    _UI,
    _MDD,
    _retnMean,
    _retnStd,
    _retnStd_div_UI,
    _CAGR,
    _CAGR_div_retnStd,
    _CAGR_div_UI,
    SPY_retnStd_d_UI,     
    SPY_CAGR,
    SPY_CAGR_d_retnStd,
    SPY_CAGR_d_UI,
  ) = symb_perf_stats_vectorized_v8(df_SPY)  

  print(f'\nSPY: retnStd/UI, CAGR/retnStd, CAGR/UI, CAGR: {SPY_retnStd_d_UI[0]:>17,.1f}, {SPY_CAGR_d_retnStd[0]:>17,.1f}, {SPY_CAGR_d_UI[0]:>17.1f}, {SPY_CAGR[0]:>17.1f}')

  # drop list with frequency count 2 or less, in l_sym_freq_cnt from zip
  zip_cnt_n_syms = zip([15,14,13,12,11,10,9,8,7,6,5,4,3], l_sym_freq_cnt[:-1])
  for item in zip_cnt_n_syms:
    sym_freq_cnt = item[0]
    syms = item[1]
    if syms:  # iterate ONLY if there are symbols in syms
      df_eval = df[iloc_start_eval:iloc_end_eval][syms]      

      if verbose:
        print(f'iloc_start_eval: {iloc_start_eval:>6},  df_eval first date: {date_start_df_eval:>10}') 
        print(f'iloc_end_eval:   {iloc_end_eval:>6},  df_eval last date:  {date_end_df_eval:>10}')        
        print(f'frequency count of symbol(s): {sym_freq_cnt}')      

        syms_n_SPY = syms + ['SPY']
        df_eval_n_SPY = df[iloc_start_eval:iloc_end_eval][syms_n_SPY]   
        print(f'\ndf_eval_n_SPY:\n{df_eval_n_SPY}\n')

      (
        _symbols,
        _period_yr,
        _retn,
        _DD,
        _UI,
        _MDD,
        _retnMean,
        _retnStd,
        _retnStd_div_UI,
        _CAGR,
        _CAGR_div_retnStd,
        _CAGR_div_UI,
        grp_retnStd_d_UI,     
        grp_CAGR,
        grp_CAGR_d_retnStd,
        grp_CAGR_d_UI,
      ) = symb_perf_stats_vectorized_v8(df_eval)  
      print(f'grp(retnStd/UI):   mean, std, mean/std: {grp_retnStd_d_UI[0]  :>17,.1f}, {grp_retnStd_d_UI[1]  :>17,.1f}, {grp_retnStd_d_UI[2]  :>17,.1f}')
      print(f'grp(CAGR/retnStd): mean, std, mean/std: {grp_CAGR_d_retnStd[0]:>17,.1f}, {grp_CAGR_d_retnStd[1]:>17,.1f}, {grp_CAGR_d_retnStd[2]:>17,.1f}')
      print(f'grp(CAGR/UI):      mean, std, mean/std: {grp_CAGR_d_UI[0]     :>17,.1f}, {grp_CAGR_d_UI[1]     :>17,.1f}, {grp_CAGR_d_UI[2]     :>17,.1f}')
      print(f'grp(CAGR):         mean, std, mean/std: {grp_CAGR[0]          :>17,.1f}, {grp_CAGR[1]          :>17,.1f}, {grp_CAGR[2]          :>17,.1f}\n')

      if store_results:  # record results to df
        row_add0      = [n_samples, str(days_lookbacks), days_eval, n_top_syms, syms_start, syms_end, sym_freq_cnt]
        row_add1      = [grp_CAGR[0],           grp_CAGR[1],           grp_CAGR[2]]
        row_add2      = [grp_CAGR_d_UI[0],      grp_CAGR_d_UI[1],      grp_CAGR_d_UI[2]]
        row_add3      = [grp_CAGR_d_retnStd[0], grp_CAGR_d_retnStd[1], grp_CAGR_d_retnStd[2]]
        row_add4      = [grp_retnStd_d_UI[0],   grp_retnStd_d_UI[1],   grp_retnStd_d_UI[2]]
        row_add5      = [SPY_CAGR[0], SPY_CAGR_d_UI[0], SPY_CAGR_d_retnStd[0], SPY_retnStd_d_UI[0]]
        row_add_total = row_add0 + row_add1 + row_add2 + row_add3 + row_add4 + row_add5
        print(f'row_add_total: {row_add_total}')






        df_eval_results.loc[len(df_eval_results)] = row_add_total
        print(f'appended row_add to df_eval_results:\n{row_add_total}\n')



        # if run_type == 'train':
        #   df_eval_sym_freq_results_train.loc[len(df_eval_sym_freq_results_train)] = row_add_total
        #   print(f'appended row_add to df_eval_sym_freq_results_train:\n{row_add_total}\n')
        # elif run_type == 'validate':
        #   df_eval_sym_freq_results_validate.loc[len(df_eval_sym_freq_results_validate)] = row_add_total       
        #   print(f'appended row_add to df_eval_sym_freq_results_validate:\n{row_add_total}\n')
        # elif run_type == 'test':
        #   df_eval_sym_freq_results_test.loc[len(df_eval_sym_freq_results_test)] = row_add_total       
        #   print(f'appended row_add to df_eval_sym_freq_results_test:\n{row_add_total}\n')
        # else:
        #   msg_stop = f"ERROR run_type must be 'train', 'validate', or 'test', run_type is: {run_type}"
        #   raise SystemExit(msg_stop )                             
  
  print('='*50, '\n')  

z_grp_top_set_syms_n_freq:
1 of 6 max_lookback_slice: (17, 137, 142)

len(idx_train): 120
idx_train: DatetimeIndex(['2022-09-06', '2022-09-07', '2022-09-08', '2022-09-09', '2022-09-12', '2022-09-13', '2022-09-14', '2022-09-15', '2022-09-16', '2022-09-19',
               ...
               '2023-02-13', '2023-02-14', '2023-02-15', '2023-02-16', '2023-02-17', '2023-02-21', '2023-02-22', '2023-02-23', '2023-02-24', '2023-02-27'], dtype='datetime64[ns]', name='Date', length=120, freq=None)
df_train_dates (inclusive): 2022-09-06 - 2023-02-27

len(idx_eval): 5
idx_eval: DatetimeIndex(['2023-02-28', '2023-03-01', '2023-03-02', '2023-03-03', '2023-03-06'], dtype='datetime64[ns]', name='Date', freq=None)
df_eval dates (inclusive): 2023-02-28 - 2023-03-06

top_set_syms_n_freq: [('FTSM', 10), ('SGEN', 9), ('TA', 8), ('ACGL', 7), ('OEC', 7), ('ACLS', 6), ('ASC', 6), ('ELF', 6), ('LNTH', 6), ('TNK', 6), ('THR', 5), ('TNP', 5), ('FTI', 4), ('HEES', 4), ('MMS', 4), ('NVGS', 4), ('PDFS', 4), ('TPX', 4

appended row_add to df_eval_results:
[6, '[15, 30, 60, 120]', 5, 20, 0, 10, 10, 0.032302244923903656, 0.0, inf, 429.6962518322571, 0.0, inf, 147.8532811817104, 0.0, inf, 2.906234128846719, 0.0, inf, 0.7522123269097252, 109.7284438363447, 65.2199475282514, 1.6824368616490148]

iloc_start_eval:    138,  df_eval first date: 2023-03-01
iloc_end_eval:      143,  df_eval last date:  2023-03-07
frequency count of symbol(s): 9

df_eval_n_SPY:
                  SGEN         SPY
Date                              
2023-03-01  178.229996  394.739990
2023-03-02  178.669998  397.809998
2023-03-03  181.289993  404.190002
2023-03-06  177.350006  404.470001
2023-03-07  177.500000  398.269989

grp(retnStd/UI):   mean, std, mean/std:               1.0,               0.0,               inf
grp(CAGR/retnStd): mean, std, mean/std:             -17.3,               0.0,              -inf
grp(CAGR/UI):      mean, std, mean/std:             -16.9,               0.0,              -inf
grp(CAGR):         mean, st

appended row_add to df_eval_results:
[6, '[15, 30, 60, 120]', 5, 20, 0, 10, 10, 0.032302244923903656, 0.0, inf, 429.6962518322571, 0.0, inf, 147.8532811817104, 0.0, inf, 2.906234128846719, 0.0, inf, 0.7522123269097252, 109.7284438363447, 65.2199475282514, 1.6824368616490148]

iloc_start_eval:    138,  df_eval first date: 2023-03-01
iloc_end_eval:      143,  df_eval last date:  2023-03-07
frequency count of symbol(s): 9

df_eval_n_SPY:
                  SGEN         SPY
Date                              
2023-03-01  178.229996  394.739990
2023-03-02  178.669998  397.809998
2023-03-03  181.289993  404.190002
2023-03-06  177.350006  404.470001
2023-03-07  177.500000  398.269989

grp(retnStd/UI):   mean, std, mean/std:               1.0,               0.0,               inf
grp(CAGR/retnStd): mean, std, mean/std:             -17.3,               0.0,              -inf
grp(CAGR/UI):      mean, std, mean/std:             -16.9,               0.0,              -inf
grp(CAGR):         mean, st

appended row_add to df_eval_results:
[6, '[15, 30, 60, 120]', 5, 20, 0, 10, 7, -0.37697106897629507, 0.4146810585994678, -0.9090626667383039, 42129.3636504454, 42192.73657778051, 0.9984980133436406, 57.27142273512827, 194.31969122927524, 0.2947278393292344, 167.80886643384824, 167.34645324153692, 1.0027632088003915, -0.575054231205185, -60.395561314041494, -90.59850658351691, 0.666628663005242]

iloc_start_eval:    127,  df_eval first date: 2023-02-13
iloc_end_eval:      132,  df_eval last date:  2023-02-17
frequency count of symbol(s): 6

df_eval_n_SPY:
                 CTLT        ELF        WYNN         SPY
Date                                                    
2023-02-13  70.550003  73.440002  109.279999  412.829987
2023-02-14  70.349998  73.769997  109.080002  412.640015
2023-02-15  69.089996  74.629997  108.459999  413.980011
2023-02-16  71.889999  74.000000  108.470001  408.279999
2023-02-17  71.370003  73.650002  109.010002  407.260010

grp(retnStd/UI):   mean, std, mean/std:

In [22]:
pickle_dump(df_eval_results, path_data_dump, fp_df_eval_results)
df = pickle_load(path_data_dump, fp_df_eval_results)
df.tail()

n_samples   days_lookbacks  days_eval  n_top_syms  syms_start  syms_end  sym_freq_cnt  grp(CAGR)_mean  grp(CAGR)_std  grp(CAGR)_mean/std  grp(CAGR/UI)_mean  grp(CAGR/UI)_std  grp(CAGR/UI)_mean/std  grp(CAGR/retnStd)_mean  grp(CAGR/retnStd)_std  grp(CAGR/retnStd)_mean/std  grp(retnStd/UI)_mean  grp(retnStd/UI)_std  grp(retnStd/UI)_mean/std  SPY_CAGR  SPY_CAGR/UI  SPY_CAGR/retnStd  SPY_retnStd/UI
800          6  [15, 30, 60,...          5          20           0        10             9       -0.933872       0.000000             -inf          -35.657905           0.000000              -inf            -108.627379                0.000000                   -inf                    0.328259              0.000000                  inf          -0.864667   -50.915082       -95.027459         0.535793
801          6  [15, 30, 60,...          5          20           0        10             8       -0.412179       0.404332        -1.019406          -21.390155          63.696110         -0.335816            -123.316093              179.229294              -0.688035                    0.978119              0.573216             1.706370          -0.864667   -50.915082       -95.027459         0.535793
802          6  [15, 30, 60,...          5          20           0        10             6       -0.573099       0.560366        -1.022724          -15.545961          23.077101         -0.673653             -19.211139               19.979990              -0.961519                    1.274893              0.451515             2.823591          -0.864667   -50.915082       -95.027459         0.535793
803          6  [15, 30, 60,...          5          20           0        10             5       -0.733702       0.302423        -2.426079          -33.247630          10.648733         -3.122215             -56.341709               24.956295              -2.257615                    0.756306              0.413644             1.828399          -0.864667   -50.915082       -95.027459         0.535793
804          6  [15, 30, 60,...          5          20           0        10             4       -0.857364       0.158130        -5.421890          -37.233417          12.800180         -2.908820             -54.547081               21.081003              -2.587499                    0.722736              0.270575             2.671113          -0.864667   -50.915082       -95.027459         0.535793

In [23]:
df = pickle_load(path_data_dump, fp_df_picks)
df.tail()

Empty DataFrame
Columns: [date_end_df_train, max_days_lookbacks, days_lookbacks, sym_freq_15, sym_freq_14, sym_freq_13, sym_freq_12, sym_freq_11, sym_freq_10, sym_freq_9, sym_freq_8, sym_freq_7, sym_freq_6, sym_freq_5, sym_freq_4, sym_freq_3, sym_freq_2]
Index: []

In [24]:
# if store_results:  # record results to df
#   if run_type == 'train':
#     pickle_dump(df_eval_sym_freq_results_train, path_data_dump, fp_df_eval_sym_freq_results_train)
#     print(f'Save results to: {fp_df_eval_sym_freq_results_train}')
#   elif run_type == 'validate':
#     pickle_dump(df_eval_sym_freq_results_validate, path_data_dump, fp_df_eval_sym_freq_results_validate)   
#     print(f'Save results to: {fp_df_eval_sym_freq_results_validate}')
#   elif run_type == 'test':
#     pickle_dump(df_eval_sym_freq_results_test, path_data_dump, fp_df_eval_sym_freq_results_test)   
#     print(f'Save results to: {fp_df_eval_sym_freq_results_test}')
#   else:
#     msg_stop = f"ERROR: Result was not saved. run_type must be 'train', 'validate', or 'test', run_type is: {run_type}"
#     raise SystemExit(msg_stop )  